GoogleDataTransport/GDTCORLibrary/GDTCOREvent.m

```diff
@@ -25,6 +25,37 @@
 
 @implementation GDTCOREvent
 
++ (NSNumber *)nextEventID {
+  static NSInteger sessionSequenceNumber = 1;
+  static unsigned long long nextEventID = 0;
+  static dispatch_queue_t eventIDQueue;
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    static NSString *kGDTCORSessionKey = @"GDTCORSession";
+    eventIDQueue = dispatch_queue_create("com.google.GDTCOREventIDQueue", DISPATCH_QUEUE_SERIAL);
+    sessionSequenceNumber = [[NSUserDefaults standardUserDefaults] integerForKey:kGDTCORSessionKey];
+    sessionSequenceNumber++;
+    [[NSUserDefaults standardUserDefaults] setInteger:sessionSequenceNumber
+                                               forKey:kGDTCORSessionKey];
+    [[NSUserDefaults standardUserDefaults] synchronize];
+  });
+
+  __block NSNumber *result;
+  dispatch_sync(eventIDQueue, ^{
+    // Concatenate the session number with the event number.
+    uint64_t tensPlace = 10;
+    while (nextEventID >= tensPlace) {
+      tensPlace *= 10;
+    }
+    uint64_t eventID = sessionSequenceNumber * tensPlace + nextEventID;
+    nextEventID++;
+    // This isn't actually concatenating numbers because the first digits aren't the
+    // sessionSequenceNumber
+    result = @(eventID);
+  });
+  return result;
```

Comment:
I wonder how important is to increment event IDs? If the only requirement for the ID is to be unique, maybe [NSUUID](https://developer.apple.com/documentation/foundation/nsuuid) will provide a simpler and more reliable implementation.

Reply:
There may be a future requirement for event IDs to be generated sequentially for determining log loss, so this will do for now. Ultimately, I may move to a double like <session counter>.<event counter>.

Reply:
I think, in order to satisfy the sequential requirement we should take care of a potential (though maybe unlikely) integer overflow. Also, curious how determining log loss will work with multiple devices sending the same ID?

Reply:
The same devices generating duplicate eventIDs will be fine. Generally, we're hoping that app instances (an period of app usage in which the cache isn't deleted) generate sequential event  IDs. The event IDs joined with some additional information an SDK generates should be able to help determine log loss. I've moved saving the counter value to a file--it seems the best way in a flat file storage system to store a value robust against crashing and lifecycle events.